## Review: Bounce-Back Cycle 2

**Reason:** CI failed (unrelated to diff)

**Reviewer A:** APPROVE — Diff correctly makes the regression test shard-deterministic via `DashMap::determine_map`; bounded search loop is convergent; load-bearing `assert_eq!` invariant guards against revert.
**Reviewer B:** APPROVE — Test-only change; parity-relevant assertions (cap behavior) unchanged. DashMap shard-collision determinism is a henyey/Rust concern with no stellar-core analogue, so no divergence introduced.
**External reviewers:** copilot-pull-request-reviewer: COMMENTED (non-blocking)
**CI:** red (unrelated)

### CI failure analysis

Failing checks are all in the `Quickstart` workflow:
- `build / 3 test / 2 test (testing-with-pr, amd64, local, rpc , )`
- `build / 3 test / 2 test (testing-with-pr, amd64, local, core,rpc,horizon , )`
- `build / 3 test / complete`
- `build / complete`

Root cause from the failed-job log: `stellar-rpc` repeatedly exits with status 1 ("supervisor: WARN exited: stellar-rpc (exit status 1; not expected)"), the friendbot test waits on `stellar-rpc` to become ready, and the action times out after 4 minutes:

```
##[error]The action 'Run friendbot test' has timed out after 4 minutes.
```

This PR's diff is **test-only and scoped to `crates/herder/`**:
- `crates/herder/Cargo.toml` (+5)
- `crates/herder/src/tx_set_tracker.rs` (+55 / -1)

It cannot affect the `stellar-rpc` Docker image's startup behavior in the Quickstart workflow — that pipeline builds and runs upstream `stellar/stellar-rpc` images independently of any henyey crate. The Rust-side checks (`Test`, `Doctest`, `Build (release-ci)`, `Clippy`, `Format`) are all green.

The failure is therefore **unrelated infrastructure flake** (or genuine upstream stellar-rpc issue on `main`).

### Action for `/do` Mode B

Rebase the branch on `origin/main` and push. No code changes required.

```bash
git fetch origin main
git rebase origin/main
git push --force-with-lease
```

If CI is still red on the same Quickstart `local, rpc` checks after the rebase with no diff-attributable cause, the next `/review-pr` will mark this `blocked` per the cycle-cap rule.

Routing back to `ready-for-doing` for `/do` Mode B.

--- a/crates/herder/Cargo.toml
+++ b/crates/herder/Cargo.toml
@@ -57,3 +57,8 @@ henyey-crypto.workspace = true
 henyey-ledger = { workspace = true, features = ["test-utils"] }
 henyey-scp = { workspace = true, features = ["test-helpers"] }
 sha2.workspace = true
+# `raw-api` exposes `DashMap::determine_map`, used by the
+# `test_request_at_cap_still_increments_existing` regression test to make the
+# shard collision between `first_hash` and `new_hash` deterministic across all
+# runners regardless of `num_cpus()`. Purely additive feature.
+dashmap = { workspace = true, features = ["raw-api"] }

--- a/crates/herder/src/tx_set_tracker.rs
+++ b/crates/herder/src/tx_set_tracker.rs
@@ -998,8 +998,62 @@ mod tests {
             "existing entry should increment even at cap"
         );
 
+        // REGRESSION GUARD for PR #2747 / issues #2744, #2750.
+        //
+        // The pre-fix code held a `dashmap::mapref::one::Ref` into `pending`
+        // across the call to `tracker.request(new_hash, ...)`. `request()`
+        // takes a write lock on the destination shard via `pending.entry(...)`.
+        // When `new_hash` and `first_hash` land on the same DashMap shard,
+        // the still-live read guard from `pending.get(&first_hash)` causes
+        // `parking_lot`'s shard `RwLock::write()` call to deadlock the same
+        // thread.
+        //
+        // DashMap's shard count is `4 * num_cpus().next_power_of_two()`, so
+        // whether two arbitrary hashes share a shard depends on the runner.
+        // To make this regression test fail _deterministically_ on every
+        // runner if the snapshot pattern above is ever reverted, we pick
+        // `new_hash` to be provably on the same shard as `first_hash` using
+        // `DashMap::determine_map` (gated behind dashmap's `raw-api`
+        // dev-feature in `crates/herder/Cargo.toml`).
+        //
+        // Inlining `tracker.pending.get(&first_hash).unwrap().request_count`
+        // into the `assert_eq!` above (or re-binding it to a `let entry =
+        // tracker.pending.get(...)` that outlives the next `tracker.request`
+        // call) resurrects the deadlock.
+        let target_shard = tracker.pending.determine_map(&first_hash);
+        let new_hash = {
+            let mut found = None;
+            for i in 0u32..1_000_000 {
+                let mut bytes = [0xFFu8; 32];
+                bytes[0..4].copy_from_slice(&i.to_le_bytes());
+                let candidate = Hash256::from_bytes(bytes);
+                if candidate == first_hash {
+                    continue;
+                }
+                if tracker.pending.contains_key(&candidate) {
+                    continue;
+                }
+                if tracker.pending.determine_map(&candidate) == target_shard {
+                    found = Some(candidate);
+                    break;
+                }
+            }
+            found.expect(
+                "must find a hash on the same DashMap shard as first_hash within 1M tries; \
+                 shard count is 4 * num_cpus().next_power_of_two() so this should converge in O(shards)",
+            )
+        };
+        // Load-bearing invariant: if this ever fires, the regression guard
+        // below has lost its teeth and the test silently stops catching the
+        // bug it was written for.
+        assert_eq!(
+            tracker.pending.determine_map(&first_hash),
+            tracker.pending.determine_map(&new_hash),
+            "test invariant: first_hash and new_hash must collide on a DashMap shard \
+             so a Ref-held-across-request() regression deadlocks deterministically"
+        );
+
         // A brand new hash should be rejected.
-        let new_hash = Hash256::from_bytes([0xFF; 32]);
         assert!(!tracker.request(new_hash, 100));
         assert!(
             tracker.pending.get(&new_hash).is_none(),

